### PR TITLE
Get JSON source and parse locally

### PR DESCRIPTION
**Pull request by imurchie**
Resurrect the old source parsing strategy (from https://github.com/appium/appium-xcuitest-driver/blob/e989e23f81bde78df3a9418f3dd10275b0009f07/lib/commands/source.js) and couple with getting the JSON source. Hide behind the `useFastSource` desired capability.

On my real device iPhone 7 Plus this makes a test which just starts a session and gets the source go from an average of 12100ms to around 7000ms.

### CODE REVIEW

File: lib/commands/source.js [hunk 1]
```diff
@@ -37,6 +42,70 @@ helpers.mobileGetSource = async function (opts = {}) {
...
+  function getTree (element, elementIndex, parentPath) {
+    let curPath = `${parentPath}/${elementIndex}`;
+    let rect = element.rect || {};
+    let subtree = {
+      '@': {
+        name: element.name || '',
+        label: element.label || '',
+        value: element.value,
+        dom: null,
+        enabled: parseInt(element.isEnabled, 10) === 1,
```
Comment (mykola-mokhnach) [inline lib/commands/source.js]: in the original xml it's either `true` or `false`

File: lib/commands/source.js [hunk 2]
```diff
@@ -37,6 +42,70 @@ helpers.mobileGetSource = async function (opts = {}) {
...
+ *     name: 'UICatalog',
+ *     rawIdentifier: null }
+ */
+function getTreeForXML (srcTree) {
+  function getTree (element, elementIndex, parentPath) {
+    let curPath = `${parentPath}/${elementIndex}`;
+    let rect = element.rect || {};
+    let subtree = {
+      '@': {
+        name: element.name || '',
```
Comment (mykola-mokhnach) [inline lib/commands/source.js]: the corresponding attribute is not shown in the original xml if the value of it is `null`

File: lib/commands/source.js [hunk 3]
```diff
@@ -37,6 +42,70 @@ helpers.mobileGetSource = async function (opts = {}) {
...
+        label: element.label || '',
+        value: element.value,
+        dom: null,
+        enabled: parseInt(element.isEnabled, 10) === 1,
+        valid: true,
+        visible: parseInt(element.isVisible, 10) === 1,
+        hint: element.hint || '',
```
Comment (mykola-mokhnach) [inline lib/commands/source.js]: `hint` attribute is not present in the xml

File: lib/commands/source.js [hunk 4]
```diff
@@ -37,6 +42,70 @@ helpers.mobileGetSource = async function (opts = {}) {
...
+        path: curPath,
```
Comment (mykola-mokhnach) [inline lib/commands/source.js]: `path` attribute is not present in the xml

**Comment by mykola-mokhnach** [comment]
Have you checked xml source performance with https://github.com/facebook/WebDriverAgent/pull/773 merged?

File: lib/commands/source.js (hunk 2 above)
Comment (mykola-mokhnach) [reply lib/commands/source.js]: the same is true for other nullable attributes below

File: README.md [hunk 5]
```diff
@@ -154,7 +154,7 @@ Differences noted here
...
 |`absoluteWebLocations`|This capability will direct the `Get Element Location` command, when used within webviews, to return coordinates which are relative to the origin of the page, rather than relative to the current scroll offset. This capability has no effect outside of webviews. Default `false`.|e.g., `true`|
 |`simulatorWindowCenter`|Allows to explicitly set the coordinates of Simulator window center for Xcode9+ SDK. This capability only has an effect if Simulator window has not been opened yet for the current session before it started.|e.g. `{-100.0,100.0}` or `{500,500}`, spaces are not allowed|
-
+|`useFastSource`|Get JSON source from WDA and parse into XML on Appium server. This can be much faster, especially on large devices. Defaults to `false`.|e.g., `true`|
```
Comment (mykola-mokhnach) [inline README.md]: I'd rename it to `useJSONSource`, since it's not the fastest available source :)

Comment (imurchie) [reply README.md]: I thought of that but that makes it seem like the source you get will be JSON.

**Comment by imurchie** [comment]
No. I'm not merging any more WDA stuff into our fork. It is too much of a pain in the ass and I don't have the time or the inclination.

File: lib/commands/source.js (hunk 1 above)
Comment (imurchie) [reply lib/commands/source.js]: This comes out in the XML result as something like
```xml
<XCUIElementTypeOther name="No SIM" label="No SIM" value="" dom="" enabled="true" valid="true" visible="true" hint="" path="/0/2/0/1/0" x="7" y="0" width="54" height="27">
```

**Comment by mykola-mokhnach** [comment]
> No. I'm not merging any more WDA stuff into our fork. It is too much of a pain in the ass and I don't have the time or the inclination.

Don't you mind if I merge the stuff myself?

**Comment by mykola-mokhnach** [comment]
Also, my point is that it's better to fix/improve issues in WDA rather than to create workarounds on our side.

**Comment by imurchie** [comment]
I understand. Nothing ever gets fixed on the WDA side, though.

You are welcome to merge things into our fork and get them into `appium-xcuitest-driver`.

**Comment by mykola-mokhnach** [comment]
> Nothing ever gets fixed on the WDA side, though.

I agree, they are slow to react. Also, we have a perfect possibility to test the changes across many platforms and get the feedback from beta users. So even if some broken stuff is merged there we can recover fast before it gets into stable.

**Pull request merged**